avr-hal-generic/src/spi.rs

```diff
@@ -290,6 +326,70 @@ where
 {
 }
 
+impl<H, SPI, SCLKPIN, MOSIPIN, MISOPIN, CSPIN> embedded_hal::spi::ErrorType
+    for Spi<H, SPI, SCLKPIN, MOSIPIN, MISOPIN, CSPIN>
+where
+    SPI: SpiOps<H, SCLKPIN, MOSIPIN, MISOPIN, CSPIN>,
+    SCLKPIN: port::PinOps,
+    MOSIPIN: port::PinOps,
+    MISOPIN: port::PinOps,
+    CSPIN: port::PinOps,
+{
+    type Error = core::convert::Infallible;
+}
+
+impl<H, SPI, SCLKPIN, MOSIPIN, MISOPIN, CSPIN> embedded_hal::spi::SpiBus
+    for Spi<H, SPI, SCLKPIN, MOSIPIN, MISOPIN, CSPIN>
+where
+    SPI: SpiOps<H, SCLKPIN, MOSIPIN, MISOPIN, CSPIN>,
+    SCLKPIN: port::PinOps,
+    MOSIPIN: port::PinOps,
+    MISOPIN: port::PinOps,
+    CSPIN: port::PinOps,
+{
+    fn read(&mut self, read: &mut [u8]) -> Result<(), Self::Error> {
+        Ok(self.p.raw_read_buf(read))
+    }
+
+    fn write(&mut self, write: &[u8]) -> Result<(), Self::Error> {
+        Ok(self.p.raw_write_buf(write))
+    }
+
+    fn transfer(&mut self, read: &mut [u8], write: &[u8]) -> Result<(), Self::Error> {
+        match read.len().cmp(&write.len()) {
+            Ordering::Less => {
+                let (write1, write2) = write.split_at(read.len());
+                self.p.raw_read_write_buf(read, write1);
+                self.p.raw_write_buf(write2);
+            }
+            Ordering::Equal => self.p.raw_read_write_buf(read, write),
+            Ordering::Greater => {
+                let (read1, read2) = read.split_at_mut(write.len());
+                self.p.raw_read_write_buf(read1, write);
+                self.p.raw_read_buf(read2);
+            }
+        }
+        Ok(())
+    }
+
+    fn transfer_in_place(&mut self, buffer: &mut [u8]) -> Result<(), Self::Error> {
+        let write = unsafe {
+            let p = buffer.as_ptr();
+            core::slice::from_raw_parts(p, buffer.len())
+        };
+        self.p.raw_read_write_buf(buffer, write);
+        Ok(())
+    }
+
+    fn flush(&mut self) -> Result<(), Self::Error> {
+        if self.write_in_progress {
```

Comment:
Just to make sure I understand this correctly:

You are only making use of the `write_in_progress` flag here to ensure correct behavior when both `e-h0.2::FullDuplex` and `e-h1.0::SpiBus` are used at the same time, right?

All the other methods of `e-h1.0::SpiBus` synchronize with the end of the transfer internally and thus they don't need to set `write_in_progress`.  Is that correct?

And related:  What if `write_in_progress` is set when calling e.g. `SpiBus::write()`?  Shouldn't `SpiBus::write()` call `<Self as SpiBus>::flush()` before operating on the bus to ensure any `FullDuplex` transfer from before has actually finished?